server/controllers/users.controller.js

```diff
@@ -1,81 +1,83 @@
-import Sequelize from 'sequelize';
-import db from '../../config/sequelize';
-
-const User = db.User;
-const Device = db.Device;
-const Op = Sequelize.Op;
+import { User } from '../models/user.model';
 
 /**
- * Start a new thread
- * @property {string} req.body.username - Body of the message
+ * Create a User
+ * @property {string} req.body.username - Username
+ * @property {string} req.body.uuid - User UUID
  * @returns {User}
  */
-function create(req, res, next) { // eslint-disable-line no-unused-vars
+export async function create(req, res) {
     const { username, uuid } = req.body;
-    User.findOrCreate({ where: { username, uuid } })
-    .spread((user, created) => { // eslint-disable-line no-unused-vars
-        res.send({ user });
-    });
-}
 
-function sendPushNotification(req, res, next) {  // eslint-disable-line no-unused-vars
-    res.send(req.body);
+    // This is the equivalent of findOrCreate
+    const user = await User.findOneAndUpdate(
+        { username, uuid },
+        { username, uuid },
+        { new: true, upsert: true }
+    );
+
+    res.send({ user });
 }
 
-function updateDevice(req, res, next) {  // eslint-disable-line no-unused-vars
+export async function updateDevice(req, res) {
     const { username, token, deviceType } = req.body;
-    User.findOne({ where: { username } }).then((deviceUser) => {
-        Device.destroy({
-            where: {
+    // delete other instances of this token
+    const otherUsersWithThisToken = await User.find({
+        username: {
+            $ne: username,
+        },
+        devices: {
+            $elemMatch: {
                 token,
                 type: deviceType,
-                UserId: {
-                    [Op.ne]: deviceUser.id,
-                },
+                status: 'enabled',
             },
-        }).then(() => {
-            Device.upsert({
-                token,
-                type: deviceType,
-                UserId: deviceUser.id,
-                disabled: null,
-            }, {
-                where: {
-                    token,
-                    type: deviceType,
-                    UserId: deviceUser.id,
-                },
-                paranoid: false,
-            })
-            .then((created) => {
-                res.send({
-                    message: `Device ${created ? 'created' : 'updated'}`,
-                });
-            });
-        });
+        },
+    });
+    for (const user of otherUsersWithThisToken) {
+        for (let i = 0; i < user.devices.length; i += 1) {
+            if (user.devices[i].token === token && user.devices[i].type === deviceType) {
+                user.devices[i].status = 'disabled';
+            }
+        }
+        await user.save();
```

Comment:
Wondering if we can do some sort of bulk update here as opposed to saving each user synchronously within each iteration.

Reply:
Doing only 1 DB operation to update a bunch of users would be nice.

Another option (and the minimum necessary, IMO) would be to kick off all the promises at once and using `Promise.all()`. Something like the following:

```js
// Don't await user.save() inside the `for..of` loop, and instead, after the `for..of` loop do
await Promise.all(otherUsersWithThisToken)
// Note: I'm not sure if a `for..of` loop modifies the original array.  My suggestion here would only work if a `for..of` loop does modify the original array.
```

Reply:
Good call! I'll look into this

Reply:
I changed this to use one big update statement